### PR TITLE
Add support for hard limit.

### PR DESCRIPTION
**Pull request by higuoxing**
This patch adds support for hard limit by @zhrt123.

### CODE REVIEW

File: quotamodel.c [hunk 1]
```diff
@@ -480,6 +481,8 @@ DiskQuotaShmemSize(void)
 	size = sizeof(ExtensionDDLMessage);
 	size = add_size(size, hash_estimate_size(MAX_DISK_QUOTA_BLACK_ENTRIES, sizeof(GlobalBlackMapEntry)));
 	size = add_size(size, hash_estimate_size(diskquota_max_active_tables, sizeof(DiskQuotaActiveTableEntry)));
+	size = add_size(size, hash_estimate_size(diskquota_max_active_tables, sizeof(DiskQuotaRelationEntry)));
+	size = add_size(size, hash_estimate_size(diskquota_max_active_tables, sizeof(DiskQuotaPgClassCacheEntry)));
```
Comment (higuoxing) [inline quotamodel.c:485]:
```suggestion
	size = add_size(size, hash_estimate_size(diskquota_max_active_tables, sizeof(DiskQuotaPgClassCacheEntry)));
	size = add_size(size, hash_estimate_size(diskquota_max_active_tables, sizeof(DiskQuotaRelidCacheEntry)));
```

It looks that we forget to allocate shared memory for `relid_cache`?

File: gp_activetable.c [hunk 2]
```diff
@@ -183,16 +429,43 @@ report_active_table_helper(const RelFileNodeBackend *relFileNode)
...
+
+			if (!found)
+			{
+				return;
+			}
 
-	if (!found && entry == NULL)
+			quota_check_common(relid_cache_entry->relid);
+			// TODO: check whether relation map already built
+			update_relation_map(relid_cache_entry->relid);
```
Comment (higuoxing) [inline gp_activetable.c:465]:
```suggestion
			quota_check_common(relid);
			// TODO: check whether relation map already built
			update_relation_map(relid);
```

The lock `diskquota_locks.pg_class_cache_lock` is released here, hence it's not safe to use `relid_cache_entry`.

File: gp_activetable.c [hunk 3]
```diff
@@ -147,7 +196,203 @@ active_table_hook_smgrtruncate(RelFileNodeBackend rnode)
...
+	{
+		relation_entry = hash_search(relation_map, &objectId, HASH_FIND, &found);
+		if(relation_entry)
+		{
+			relation_entry->relfilenode = relfileNode.relNode;
+		}
+	}
+	LWLockRelease(diskquota_locks.pg_class_cache_lock);
+
+	report_active_table_helper(&relfileNode);
```
Comment (higuoxing) [inline gp_activetable.c:312]: We are able to construct `RelFileNodeBackend` by the following code snippet.

```C
RelFileNodeBackend relfilenode;
relfilenode.node = rel->rd_node;
relfilenode.backend = rel->rd_backend;
```

No need to modify the interface of `report_active_table_helper()`. It's good to keep change smaller.

**Review by higuoxing — COMMENTED** [review]
Couple of comments from myself. I'm still working on resolving failing tests.

File: gp_activetable.c (hunk 2 above)
Comment (zhrt123) [reply gp_activetable.c:465]: The relid maybe INVALID, so we cannot use it to check quota limit.

File: quotamodel.c (hunk 1 above)
Comment (zhrt123) [reply quotamodel.c:485]: Yes, it is my mistake.

File: gp_activetable.h [hunk 4]
```diff
@@ -10,7 +10,20 @@ typedef struct DiskQuotaActiveTableFileEntry
...
+	Oid			relfilenode;
+	Oid			relid;
+}	DiskQuotaRelidCacheEntry;
+
+typedef struct DiskQuotaRelationEntry
+{
+	Oid			relid;
+	Oid			relfilenode;
+	Oid			primary_table_oid;
+}	DiskQuotaRelationEntry;
```
Comment (higuoxing) [inline gp_activetable.h:26]:
```
typedef struct RelFileNode
{
	Oid			spcNode;		/* tablespace */
	Oid			dbNode;			/* database */
	Oid			relNode;		/* relation */
} RelFileNode;
```

**Comment by higuoxing** [comment]
Close, in favor of #80.

**Pull request closed without merging**